docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1356 @@
+{
```

Comment:
<p>I've thought about this more and this is not going to work for production code (i.e., pre-computing z-normalized subsequences is not scalable). For something as short as <code>n = 1_000_000</code>:</p><p> </p><pre class="ql-syntax" spellcheck="false">import numpy as np
from stumpy import core

T = np.random.rand(1_000_000)
m = 50

subseqs = core.z_norm(core.rolling_window(T, 1000), axis=1) 
</pre><p> </p><p>Required over 12GB of RAM to process (due to intermediate arrays) and the final <code>subseqs</code>  required 7GB of RAM to store. For <code>n = 10_000_000</code>,  this took way more RAM than my laptop could provide. This means that we'll need to accept a slower computational speed in exchange for memory scalability. All of our functions need to be changed to take <code>T</code>  and <code>m</code>  as inputs and we'll need to normalize them on the fly.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>

Reply:
That's right. Since we are not going to focus on MERLIN3 in this PR, I will do z-normalization on the fly.
---

There is still one problem. If the timeseries gets very very long, we still need to z-normalize ALL so-far-selected candidates (see `np.matmul()` in the line 52 of `_select_candidates` function). This can become a problem we have discovered a lot of candidates.

(But, I assume we are going to break that down into batches  to resolve  the memory issue in that case)

Reply:
<p>Can you tell me if <code>left_non_trivial_cand_idx</code>  is always a set of continuous indices (i.e., the indices are increasing monotonically without skipping)? If the answer is <code>yes</code>  then we should just replace <code>np.matmul</code>  with <code>core._mass</code>  instead and calculate the actual distance. It will be slightly slower but:</p><p> </p><ol><li>The implementation will be much closer to the published work</li><li>The implementation will be scalable to much larger arrays</li></ol>

Reply:
> Can you tell me if `left_non_trivial_cand_idx` is always a set of continuous indices (i.e., the indices are increasing monotonically without skipping)?

I don' think so (a simple example: if we have two clear anomaly in the time series). So, we "might" have slices of continuous indices. 

I will do an investigation and will get back to you.

Reply:
<p>sounds good</p>

Reply:
> Can you tell me if `left_non_trivial_cand_idx` is always a set of continuous indices (i.e., the indices are increasing monotonically without skipping)?

I tracked the changes of `left_non_trivial_cand_idx` through the iterations in the function `_select_candidates`. As we expected, it is **NOT** a set of continuous indices. The indices are increasing monotonically; HOWEVER, there exists skipping (gaps)! 

Below is an example of `left_non_trivial_cand_idx` observed in one of the iterations in function `_select_candidates`.
`[57 58 59 73 74 80 81 82 83]`.

---

**Solution (?)**
IF we change the  `for-loop` of function `_select_candidates`  and, instead, iterate block-by-block (similar to MERLIN3 MATLAB code). Then, we can titerate through the candidates  (contained by `left_non_trivial_cand_idx`) one by one. Then, we can use STUMPY `mass_absolute` (where, `Q` is one of the candidate,  and `T` is the timeseries in that block. (just a thought to consider in the future)

Reply:
@Ninimama For `left_non_trivial_cand_idx = np.array([57, 58, 59, 73, 74, 80, 81, 82, 83])`, we can use:

```
import numpy as np

def _get_chunks_ranges(a):
    repeats = np.full(len(a), 2)
    diff_is_one = np.diff(a) == 1
    repeats[1:] -= diff_is_one
    repeats[:-1] -= diff_is_one
    out = np.repeat(a, repeats).reshape(-1, 2)
    out[:, 1] += 1

    return out

chunks = _get_chunks_ranges(left_non_trivial_cand_idx)
```

which will produce the (inclusive) start index and the (exclusive) stop index:

```
array([[57, 60],
       [73, 75],
       [80, 84]])
```

And then you can loop across each row and compute the distances. Would you like to try implementing this along with using `_mass` (`_mass_absolute` is for non-normalized methods)?

Reply:
Good suggestion. Even if we have narrow-range slice in chunks, this should help us. 

By the way, we already have something similar implemented in `stumpy.core` :  see `_get_mask_slices(mask)`. 
So, instead of:
```
left_non_trivial_cand_idx = np.flatnonzero(is_cands[:i-excl_zone])
```

I can do:
```
chunks = stumpy.core._get_mask_slices(is_cands[:i-excl_zone])
```



> Would you like to try implementing this along with using `_mass` (`_mass_absolute` is for non-normalized methods)?

Sure. I will try to implement it in the next commits.

Reply:
<p>Reusing existing code? Even better!!</p>